data-upgrade-users/src/main/java/org/exoplatform/migration/UserPasswordHashMigration.java

```diff
@@ -0,0 +1,142 @@
+    public void processUpgrade(String s, String s1) {
+        LOG.info("Start upgrade of users passwords hashing algorithm");
+        long startupTime = System.currentTimeMillis();
+
+        PortalContainer container = PortalContainer.getInstance();
+        RequestLifeCycle.begin(container);
+        AtomicInteger updatedPasswords = new AtomicInteger();
+        EntityManager entityManager = this.entityManagerService.getEntityManager();
+        try {
+            String sqlString = "SELECT jbid_io.NAME, jbid_io_creden.TEXT FROM " +
+                    " ((jbid_io_attr INNER JOIN jbid_io_creden ON jbid_io_attr.IDENTITY_OBJECT_ID = jbid_io_creden.IDENTITY_OBJECT_ID)" +
+                    " INNER JOIN jbid_io ON jbid_io_attr.IDENTITY_OBJECT_ID = jbid_io.ID)" +
+                    " INNER JOIN( SELECT jbid_io_attr.IDENTITY_OBJECT_ID," +
+                    " min(CASE WHEN jbid_io_attr.NAME  = 'passwordSalt' THEN jbid_io_attr.NAME  ELSE NULL END) AS salt," +
+                    " min(CASE WHEN jbid_io_attr.NAME = 'passwordSalt128' THEN jbid_io_attr.NAME  ELSE NULL END) AS salt128" +
+                    " FROM jbid_io_attr  GROUP BY jbid_io_attr.IDENTITY_OBJECT_ID" +
+                    " HAVING salt IS NOT NULL AND salt128 IS NULL) jia ON jbid_io_attr.IDENTITY_OBJECT_ID  = jia.IDENTITY_OBJECT_ID;";
+
+            Query nativeQuery = entityManager.createNativeQuery(sqlString);
+            List<Object[]> result = nativeQuery.getResultList();
+            result.forEach(item -> {
+                String userName = (String) item[0];
+                String passwordHash = (String) item[1];
+                try {
+                    String saltString = Hex.encodeHexString(generateRandomSalt());
+                    User user = picketLinkIDMService.getIdentitySession().getPersistenceManager().findUser(userName);
+                    picketLinkIDMService.getExtendedAttributeManager().addAttribute(userName, PASSWORD_SALT_USER_ATTRIBUTE, saltString);
+                    picketLinkIDMService.getExtendedAttributeManager().updatePassword(user, passwordHash);
+                    updatedPasswords.getAndIncrement();
+                    LOG.info("{}/{} passwords have been updated", updatedPasswords.get(), result.size());
+                } catch (Exception e) {
+                    LOG.error("Error while creating attribute salt and updating password hash for user : {}", userName, e);
+                }
+            });
+        } catch (Exception e) {
+            LOG.error("Error while getting old users passwords hash", e);
+        } finally {
+            RequestLifeCycle.end();
+        }
+        LOG.info("End upgrade of users passwords hashing algorithm. {} passwords has been updated. It took {} ms",
+                updatedPasswords.get(),
+                (System.currentTimeMillis() - startupTime));
+    }
+
+    @Override
+    public void afterUpgrade() {
+        settingService.set(Context.GLOBAL.id(USER_PASSWORD_HASH_MIGRATION),
```

Comment:
This should be set only if success.

In your case, if there are 10k users to update, and if I stop the server before the end, it will set it as finished. but it is not

Reply:
I didn't understand, if you stop the server this will not be called, thus i added it

Reply:
I thought that afterUpgrade is called, even the upgrade is not finished or failed.

Other question : imagine that during the execution, there is a problem for some users (not all). 

Then the execution finished.
Does the migration is considered as finished or not ?
Does it will restart at next startup ?

In addition, you configure upgradePlugin as executed once, so why do we need to store a setting to know if the execution is finished ?

In fact, in other upgrade plugin, there are 2 option

- UP finish with an exeception : so it is considered as not executed and will relaunched at next startup
- UP finish without exception : so it is considered as executed, and will not relaunched at next startup
With this assumption, do we really need a setting to mark it as finished ?

Reply:
no sure. Need to speak about that. Can you ping me ?